src/main/java/org/opensearch/ad/ratelimit/ConcurrentQueue.java

```diff
@@ -0,0 +1,162 @@
+     * @param random Random number generator
+     * @param adCircuitBreakerService AD Circuit breaker service
+     * @param threadPool threadpool accessor
+     * @param settings Cluster settings getter
+     * @param maxQueuedTaskRatio maximum queued tasks ratio in ES threadpools
+     * @param clock Clock to get current time
+     * @param mediumSegmentPruneRatio the percent of medium priority requests to prune when the queue is full
+     * @param lowSegmentPruneRatio the percent of low priority requests to prune when the queue is full
+     * @param maintenanceFreqConstant a constant help define the frequency of maintenance.  We cannot do
+     *   the expensive maintenance too often.
+     * @param clientUtil utility with ES client
+     * @param concurrencySetting Max concurrent processing of the queued events
+     * @param executionTtl Max execution time of a single request
+     * @param stateTtl max idle state duration.  Used to clean unused states.
+     * @param nodeStateManager node state accessor
+     */
+    public ConcurrentQueue(
+        String queueName,
+        long heapSizeInBytes,
+        int singleRequestSizeInBytes,
+        Setting<Float> maxHeapPercentForQueueSetting,
+        ClusterService clusterService,
+        Random random,
+        ADCircuitBreakerService adCircuitBreakerService,
+        ThreadPool threadPool,
+        Settings settings,
+        float maxQueuedTaskRatio,
+        Clock clock,
+        float mediumSegmentPruneRatio,
+        float lowSegmentPruneRatio,
+        int maintenanceFreqConstant,
+        ClientUtil clientUtil,
+        Setting<Integer> concurrencySetting,
+        Duration executionTtl,
+        Duration stateTtl,
+        NodeStateManager nodeStateManager
+    ) {
+        super(
+            queueName,
+            heapSizeInBytes,
+            singleRequestSizeInBytes,
+            maxHeapPercentForQueueSetting,
+            clusterService,
+            random,
+            adCircuitBreakerService,
+            threadPool,
+            settings,
+            maxQueuedTaskRatio,
+            clock,
+            mediumSegmentPruneRatio,
+            lowSegmentPruneRatio,
+            maintenanceFreqConstant,
+            stateTtl,
+            nodeStateManager
+        );
+
+        this.permits = new Semaphore(concurrencySetting.get(settings));
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(concurrencySetting, it -> permits = new Semaphore(it));
+        this.clientUtil = clientUtil;
+
+        this.lastExecuteTime = clock.instant();
+        this.executionTtl = executionTtl;
+    }
+
+    @Override
+    public void maintenance() {
+        super.maintenance();
+
+        if (lastExecuteTime.plus(executionTtl).isBefore(clock.instant())) {
+            LOG.warn("previous execution has been running for too long.  Maybe there are bugs.");
+
+            // release one permit
+            permits.release();
```

Comment:
Why do we release one permit here?

Reply:
Please check the comment at line 127.

Reply:
Right, but why 1 permit and not multiple?

Reply:
because one request uses one permit.  After the request finished, we need to release one.